### ISSUE TITLE
Has WET Tables (using datatables.net) been assessed for WCAG, specifically 3.2.2?

### ISSUE THREAD

**Issue by alanbec** [wet-boew/wet-boew#8434]
We are considering following [WET Tables](https://wet-boew.github.io/v4.0-ci/demos/tables/tables-en.html) example (powered by datatables.net) to provide, click-to-sort table headers and searchable tables.

Concerning accessibility, it looked like perhaps [WCAG 3.2.2](https://www.w3.org/TR/WCAG21/#on-input) might be failed since meeting it requires:

> Changing the setting of any user interface component does not automatically cause a [change of context](https://www.w3.org/TR/WCAG21/#dfn-change-of-context) unless the user has been advised of the behavior before using the component.

Assuming WET Tables were assessed for WCAG, who might I ask re their view on 3.2.2 - I am not sure the operation of the sortable headers for example actually _does_ contravene 3.2.2, and I would value the opinion of others.

Cheers, -Alan

**Comment by juleskuehn** [comment]
Reordering the table content at the user's request is not an example of a change in context, any more than filtering the content via the text input is a change in context (which it is not):

> A change of content is not always a change of context. **This success criterion is automatically met if changes in content are not also changes of context.**

https://www.w3.org/WAI/WCAG21/Understanding/on-input.html

> A change of content is not always a change of context. **Changes in content, such as an expanding outline, dynamic menu, or a tab control do not necessarily change the context**, unless they also change one of the above (e.g., focus). 

https://www.w3.org/TR/WCAG21/#dfn-change-of-context

**Comment by alanbec** [comment]
Thank you.

Yes:

> A change of content is not always a change of context. **Changes in content, such as an expanding outline, dynamic menu, or a tab control do not necessarily change the context**, unless they also change one of the above (e.g., focus).
>https://www.w3.org/TR/WCAG21/#dfn-change-of-context

However, notwithstanding that, might clicking a heading to sort a table, or, using the (dynamic) 'Search' facility* (*admittedly not the subject of this thread) be deemed as meeting the criteria for _change of context_ by possibly being classed as an operation that was:

> ...significantly re-arranging the content of a page... (https://www.w3.org/TR/WCAG21/#dfn-change-of-context)

or is such a change to the content of a page not significant _enough_, or too localized to meet that criteria?

I was looking for a tighter definition of _'significantly re-arranging the content of a page'_ to try and definitively answer my own question, but failing to find one I opted to ask here -- thanks in advance for your comments.

**Comment by juleskuehn** [comment]
There is a second reason why these controls meet criterion 3.2.2, which is that the user is expecting the behavior as it is explicitly indicated both visually and by aria labels.

More quotes pulled from https://www.w3.org/WAI/WCAG21/Understanding/on-input.html :

> Success Criterion 3.2.2 On Input (Level A): Changing the setting of any user interface component does not automatically cause a change of context unless the **user has been advised of the behavior before using the component.** 

> Changes of context are appropriate only when it is **clear that such a change will happen in response to the user's action.** 

> This Success Criterion covers changes in context due to changing the setting of a control. **Clicking on links or tabs in a tab control is activating the control, not changing the setting of that control.** 

Success example: 

> A form contains fields representing US phone numbers. All of the numbers have a three digit area code followed by a three digit prefix and finally a four digit number, and each part of the phone number is entered into a separate field. When the user completes the entry of one field the focus automatically moves to the next field of the phone number. **This behavior of phone fields is described for the user at the beginning of the form. **

**Comment by juleskuehn** [comment]
You may also wish to see the failure examples on that same page. The main failures relate to a situation where changing a form element automatically submits a form, opens a new window, or otherwise changes the web page substantially (such as redirecting to a different page).

Thanks for looking into it Alan.

**Comment by alanbec** [comment]
Welcome @juleskuehn and thanks so much for your comments, I was not taking into account the ARIA added by datatables.

**Comment by alanbec** [comment]
[I hope this is sufficiently related and so reopening here is OK] Related: Re: the search box, an 'X' (to cancel search) is added if a search is carried, but if the input looses focus the 'X' disappears only to reappear on hover, so this hide/appear only works for desktop not mobile OS users. This seems like a reduction in accessibility compared to the 'X' being visible at all times until clicked. Has anyone used the table in a way where the 'X' is persistent or some other way to avoiding this arguable dip in accessibility?

**Comment by juleskuehn** [comment]
Could you clarify what you mean, perhaps with screenshots? I can't make an "X" appear.

**Comment by alanbec** [comment]
Happily @juleskuehn 

One I have searched for 'Vanc' and my mouse is over the search box, the other I have moused away

![2018-08-24_16-23-37](https://user-images.githubusercontent.com/41155257/44606183-333a1400-a7ba-11e8-8394-529f5e9c39d3.png)

![2018-08-24_16-23-04](https://user-images.githubusercontent.com/41155257/44606182-333a1400-a7ba-11e8-92c1-88b0bfdade5b.png)

If I mouse over the 'X' reappears.

**Comment by thekodester** [comment]
I believe that is the web browser generating the 'X' as if I test:

http://wet-boew.github.io/v4.0-ci/demos/tables/tables-en.html
https://datatables.net/examples/api/multi_filter.html

I do not get the 'X' on some browsers.

**Comment by juleskuehn** [comment]
On what page? I believe this is an enhancement made by the browser, unrelated to DataTables or WET, and, as it is merely an enhancement it is not an accessibility issue.

**Comment by grifare** [comment]
There is no issue here.
Therefore, there is no accessibility issue here as well.
That cross sign, **inserted by IE**, is a feature IE provides for deleting the search entry. Its usefulness is highly debatable, but the website owners are not supposed to overwrite each and every browser feature.

Accessibility issues are not hidden in tiny little details: @duboisp this issue should be closed.

I would like to provide examples from major websites, 
[https://www.gov.uk/](url)
![image](https://user-images.githubusercontent.com/3210595/70085599-90885c00-15de-11ea-8378-a4694f9f2aa3.png)
[(https://www.usa.gov/)](url)
![image](https://user-images.githubusercontent.com/3210595/70085757-e8bf5e00-15de-11ea-9072-eaccfaab4230.png)
[http://www.w3.org/](url)
![image](https://user-images.githubusercontent.com/3210595/70086134-b6fac700-15df-11ea-9304-d7ddbacf2f45.png)
